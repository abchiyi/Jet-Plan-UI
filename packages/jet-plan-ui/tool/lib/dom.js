export function getOffset (el) {
    function getWindowScroll () {
        // FullSupport
        const supportPageOffset = window.pageXOffset !== undefined
        const isCSS1Compat = ((document.compatMode || "") === "CSS1Compat")

        const x = supportPageOffset ?
            window.pageXOffset : isCSS1Compat ?
                document.documentElement.scrollLeft :
                document.body.scrollLeft

        const y = supportPageOffset ?
            window.pageYOffset : isCSS1Compat ?
                document.documentElement.scrollTop :
                document.body.scrollTop

        return {
            x: x,
            y: y
        }
    }

    // 获取元素内部不含边框相对页面的实际距离
    const elScreenBounding = el.getBoundingClientRect()
    const windowScrool = getWindowScroll()
    const scrollTop = windowScrool.y || el.scrollTop
    const scrollLeft = windowScrool.x || el.scrollLeft

    const elWidth = elScreenBounding.right - elScreenBounding.left
    const elHeight = elScreenBounding.bottom - elScreenBounding.top
    return {
        size: elScreenBounding,
        // XXX不适合的属性名称
        elWidth: elWidth,
        elHeight: elHeight,
        y: elScreenBounding.top + scrollTop - el.clientTop,
        x: elScreenBounding.left + scrollLeft - el.clientLeft
    }
}

export function shadowPainter (direction, intensity, color) {



    function calcDirection (array) {
        if (array.indexOf(direction) !== -1) {
            return direction == array[1] ?
                intensity + "px" :
                intensity * -1 + 'px'
        }
        return '0'
    }

    function x () {
        return calcDirection(['left', 'right'])
    }

    function y () {
        return calcDirection(['top', 'bottom'])
    }
    chekcDirection(direction)
    return `
    ${x()} ${y()}
    ${1.5 * intensity}px
    ${0.1 * intensity}px
    ${color ? color : 'var(--material-shadow)'}`
}
export class SHADOW_PAINTER_ERROR extends Error {
    constructor (message) {
        super(message)
        this.name = 'ShadowPainterError'
    }
}

export function chekcDirection (direction) {
    const args = ['center', 'top', 'bottom', 'left', 'right']
    if (direction) {
        if (args.indexOf(direction.toLocaleLowerCase()) == -1) {
            throw new SHADOW_PAINTER_ERROR(
                `ShadowPainter:DIRECTON_ERROR:ErrorArg: ${direction}
                Only supports:
                [default:center ['top', 'bottom', 'left', 'right']]`)
        }
    }

}
