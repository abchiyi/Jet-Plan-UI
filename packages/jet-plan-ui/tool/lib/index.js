export class TimedActionLimit {
    constructor(timeout, limit = 1, waitCoolingDown = true) {
        if (!timeout) {
            throw SyntaxError(
                `
Class TimedActionLimit([timeout,[limit,waitCoolingDown]])
Missing Value "timeout"
Missing values ‘timeout’ are required
`
            );
        }
        this.waitCoolingDown = waitCoolingDown;
        this.limit = limit;
        this.timeout = timeout;
        this.timeoutID = [];
        this.overheat = false;
        this.conter = limit;
        this.cooledAlarm = false;
        this.overheatAlarm = false;
    }

    action(callback) {
        if (!this.overheat) {
            this.conter = this.conter - 1;
            this.overheat = this.conter == 0;
            callback();
        }
        if (this.overheat) {
            // 通知外部程序已过热
            if (this.overheatAlarm) this.overheatAlarm();
            // 设置冷却动作
            this.__setCooling();
            // 清除冷却动作
            this.__waitControl();
        }
    }

    __setCooling() {
        // 设置冷却动作
        this.timeoutID.push(
            setTimeout(() => {
                this.overheat = false;
                // 通知外部程序已冷却
                if (this.cooledAlarm) this.cooledAlarm();
                this.timeoutID = [];
            }, this.timeout)
        );
        this.conter = this.limit;
    }

    __waitControl() {
        if (this.timeoutID.length > 1) {
            const clId = this.waitCoolingDown
                ? this.timeoutID.shift()
                : this.timeoutID.pop();
            clearTimeout(clId);
        }
    }

    setCooledAlarm(callback) {
        this.cooledAlarm = callback;
    }

    setOverheatAlarm(callback) {
        this.overheatAlarm = callback;
    }
}

export function installComponent(Vue, array) {
    array.forEach((component) => {
        Vue.component(component.name, component);
    });
}
export function installDirective(Vue, array) {
    array.forEach((directive) => {
        Vue.directive(directive.name, directive);
    });
}

export function hexToRgb(hex) {
    const hashtag = hex[0] == "#";
    function parse(v) {
        return parseInt(`0x${v}`);
    }
    let rgb = {
        r: hashtag ? parse(hex.slice(1, 3)) : parse(hex.slice(0, 2)),
        g: hashtag ? parse(hex.slice(3, 5)) : parse(hex.slice(2, 4)),
        b: hashtag ? parse(hex.slice(5, 7)) : parse(hex.slice(4, 6)),
        opacity:
            parseFloat(
                (hashtag ? parse(hex.slice(7, 9)) : parse(hex.slice(6, 8))) /
                    255
            ).toFixed(2) + 5,
    };
    // let max = rgb.r + rgb.g + rgb.b;
    return {
        ...rgb,
        max: Math.abs(
            parseFloat(
                (parseInt(rgb.r) + parseInt(rgb.g) + parseInt(rgb.b)) /
                    (255 * 3)
            ).toFixed(3)
        ),
        color:
            rgb.opacity < 1
                ? `rgba(${rgb.r},${rgb.g},${rgb.b},${rgb.opacity})`
                : `rgb(${rgb.r},${rgb.g},${rgb.b})`,
    };
}

export function rgbToHex(rgb) {
    function colorToHex(color) {
        var hexadecimal = color.toString(16);
        return hexadecimal.length == 1 ? "0" + hexadecimal : hexadecimal;
    }
    const leftBrackets = rgb.indexOf("(") + 1;
    const rightBrackets = rgb.indexOf(")");
    let rgba = rgb.substring(leftBrackets, rightBrackets).split(",");
    let s = ["#"];
    let conter = 0;
    rgba.forEach((v) => {
        conter++;
        v = v.replace(/^\s+|\s+$/gm, "");
        s.push(
            conter < 4 ? colorToHex(Math.abs(v)) : colorToHex(parseInt(v * 255))
        );
    });
    return {
        color: s.join(""),
        max: Math.abs(
            parseFloat(
                (parseInt(rgba[0]) + parseInt(rgba[1]) + parseInt(rgba[2])) /
                    (255 * 3)
            ).toFixed(3)
        ),
    };
}
