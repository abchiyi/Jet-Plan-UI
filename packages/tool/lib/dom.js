import $ from 'jquery';
import 'jquery.easing'


export function getOffset (el) {
    function getWindowScroll () {
        // FullSupport
        const supportPageOffset = window.pageXOffset !== undefined;
        const isCSS1Compat = ((document.compatMode || "") === "CSS1Compat");

        const x = supportPageOffset ?
            window.pageXOffset : isCSS1Compat ?
                document.documentElement.scrollLeft :
                document.body.scrollLeft;

        const y = supportPageOffset ?
            window.pageYOffset : isCSS1Compat ?
                document.documentElement.scrollTop :
                document.body.scrollTop;

        return {
            x: x,
            y: y
        }
    }

    // 获取元素内部不含边框相对页面的实际距离
    const elScreenBounding = el.getBoundingClientRect()
    const windowScrool = getWindowScroll()
    const scrollTop = windowScrool.y || el.scrollTop
    const scrollLeft = windowScrool.x || el.scrollLeft

    return {
        size: elScreenBounding,
        // XXX不适合的属性名称
        y: elScreenBounding.top + scrollTop - el.clientTop,
        x: elScreenBounding.left + scrollLeft - el.clientLeft
    }
}

export function scrollTo (elID, toTop) {
    let position;
    if (elID.toLowerCase() === "top") {
        // 保留字段,返回顶部
        position = 0
    } else {
        const SCREENHEIGHT = getOffset($('body')[0]).size.height
        const ELHEIGHT = getOffset($(elID)[0]).size.height
        const TOP = $(elID).offset().top

        // 设置 toTop 的默认值
        // FIXME不需要的默认值，容易引发逻辑错误
        toTop = toTop ? Math.round(toTop) : 200

        if (ELHEIGHT <= SCREENHEIGHT) {
            // 当元素接近屏幕高度时
            position = ELHEIGHT >= SCREENHEIGHT * 0.75 ?
                TOP - toTop :
                TOP - (SCREENHEIGHT - ELHEIGHT) / 2
        } else {
            position = TOP - toTop
        }

        position = position < 1 ? 0 : position // 元素在顶部时,设置为0
    }


    $('html, body').animate({
        scrollTop: position
    }, {
        // TODO 使用UI 定义的缓动函数,和根据元素滚动距离定义的动画时间
        duration: 500,
        easing: 'easeInOutQuad'
    })
    return false

}

export function shadowPainter (direction, intensity, color) {



    function calcDirection (array) {
        if (array.indexOf(direction) !== -1) {
            return direction == array[1] ?
                intensity + "px" :
                intensity * -1 + 'px'
        }
        return '0'
    }

    function x () {
        return calcDirection(['left', 'right'])
    }

    function y () {
        return calcDirection(['top', 'bottom'])
    }
    chekcDirection(direction)
    return `
    ${x()} ${y()}
    ${1.5 * intensity}px
    ${0.3 * intensity}px
    ${color ? color : 'var(--shadow)'}`
}


export class SHADOW_PAINTER_ERROR extends Error {
    constructor (message) {
        super(message);
        this.name = 'ShadowPainterError';
    }
}

export function chekcDirection (direction) {
    const args = ['center', 'top', 'bottom', 'left', 'right']
    if (direction) {
        if (args.indexOf(direction.toLocaleLowerCase()) == -1) {
            throw new SHADOW_PAINTER_ERROR(
                `ShadowPainter:DIRECTON_ERROR:ErrorArg: ${direction}
                Only supports:
                [default:center ['top', 'bottom', 'left', 'right']]`)
        }
    }

}
