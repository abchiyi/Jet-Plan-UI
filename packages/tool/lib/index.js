export class TimedActionLimit {
    constructor (timeout, limit = 1) {
        this.limit = limit;
        this.timeout = timeout;
        this.timeoutID = [];
    }

    action (callback) {
        const timeoutObj = setTimeout(
            () => {
                callback()
            }, this.timeout
        )
        this.timeoutID.push(timeoutObj)

        // 移除超时的动作
        setTimeout(
            () => {
                let index = this.timeoutID.indexOf(timeoutObj)
                if (index != -1) this.timeoutID.splice(index, 1)
            }, this.timeout
        )

        // 多次时间内执行动作,取消执行先前动作
        if (this.timeoutID.length > this.limit) {
            let id = this.timeoutID.shift()
            clearTimeout(id)
        }
    }
}

export function installComponent (Vue, array) {
    array.forEach(component => {
        Vue.component(component.name, component)
    })
}
export function installDirective (Vue, array) {
    array.forEach(directive => {
        Vue.directive(directive.name, directive)
    })
}
