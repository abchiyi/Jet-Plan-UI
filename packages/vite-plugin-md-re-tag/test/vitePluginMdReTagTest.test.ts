import { describe, test, expect } from "vitest";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import fs from "fs";

import { replaceCodeBox, genToc, injectHTagId } from "../lib";

function readeMdFile(path: string): string {
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = dirname(__filename);

  return fs.readFileSync(resolve(__dirname, path)).toString();
}

const testCode = readeMdFile("./test.md");
const testCode2 = readeMdFile("./test2.md");
const testCode3 = readeMdFile("./test3.md");

describe("vitePluginMdReTagTest", () => {
  test("replaceCodeBox", () => {
    const code = replaceCodeBox(testCode);
    expect(code).toMatchSnapshot();
  });
  test("从 .md 文本生成 Toc 数据", () => {
    expect(genToc(testCode)).toMatchSnapshot();
    expect(genToc(testCode2)).toMatchSnapshot();
    expect(genToc(testCode3)).toMatchSnapshot();
    expect(genToc(testCode3)).toEqual([
      {
        children: [
          { children: [], level: 2, text: "安装组件库" },
          { children: [], level: 2, text: "注册组件" },
        ],
        level: 1,
        text: "快速开始",
      },
    ]);
  });
  test("为 'h' 标签生成锚点", () => {
    expect(injectHTagId(testCode)).toMatchSnapshot();
    expect(injectHTagId(testCode2)).toMatchSnapshot();
    expect(injectHTagId(testCode3)).toMatchSnapshot();
  });
});
