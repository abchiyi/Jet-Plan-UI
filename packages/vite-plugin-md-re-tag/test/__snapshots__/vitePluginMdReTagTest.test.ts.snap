// Vitest Snapshot v1

exports[`vitePluginMdReTagTest > replaceCodeBox 1`] = `
"# h1

## h2

### h3

## h2-2

### h3-2

#### h4

##### h5

###### h6

<script setup lang=\\"ts\\">
import { Transition } from \\"vue\\";
</script>

\`\`\`xml
<script>
    import { defineComponent, h } from \\"vue\\";
    import { theme } from 'theme.js'
    export default defineComponent({
        setup(){
            return {
                background: theme.toMap().background,
                theme,
            }
        },
        render(
            return h('div',{
                class:\\"demo\\",
                style: {
                    color: theme.toMap().text.light， // 应用 css 变量
                }
            })
        )
    })
</script>
<style>
:root{
    background:red
}
</style>
\`\`\`

<!-- 缩进测试 -->
- ## 标题

    \`\`\`ts
    console.log('test code')
    \`\`\`
"
`;

exports[`vitePluginMdReTagTest > 为 'h' 标签生成锚点 1`] = `
"<h1 id=\\"h1\\">h1</h1>

<h2 id=\\"h2\\">h2</h2>

<h3 id=\\"h3\\">h3</h3>

<h2 id=\\"h2-2\\">h2-2</h2>

<h3 id=\\"h3-2\\">h3-2</h3>

<h4 id=\\"h4\\">h4</h4>

<h5 id=\\"h5\\">h5</h5>

<h6 id=\\"h6\\">h6</h6>

<script setup lang=\\"ts\\">
import { Transition } from \\"vue\\";
</script>

\`\`\`xml
<script>
    import { defineComponent, h } from \\"vue\\";
    import { theme } from 'theme.js'
    export default defineComponent({
        setup(){
            return {
                background: theme.toMap().background,
                theme,
            }
        },
        render(
            return h('div',{
                class:\\"demo\\",
                style: {
                    color: theme.toMap().text.light， // 应用 css 变量
                }
            })
        )
    })
</script>
<style>
:root{
    background:red
}
</style>
\`\`\`

<!-- 缩进测试 -->
- ## 标题

    \`\`\`ts
    console.log('test code')
    \`\`\`
"
`;

exports[`vitePluginMdReTagTest > 为 'h' 标签生成锚点 2`] = `
"test page
测试在没有标题的情况
"
`;

exports[`vitePluginMdReTagTest > 为 'h' 标签生成锚点 3`] = `
"<h1 id=\\"快速开始\\">快速开始</h1>

<h2 id=\\"安装组件库\\">安装组件库</h2>

<h2 id=\\"注册组件\\">注册组件</h2>
"
`;

exports[`vitePluginMdReTagTest > 从 .md 文本生成 Toc 数据 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "level": 3,
            "text": "h3",
          },
        ],
        "level": 2,
        "text": "h2",
      },
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "level": 6,
                        "text": "h6",
                      },
                    ],
                    "level": 5,
                    "text": "h5",
                  },
                ],
                "level": 4,
                "text": "h4",
              },
            ],
            "level": 3,
            "text": "h3-2",
          },
        ],
        "level": 2,
        "text": "h2-2",
      },
    ],
    "level": 1,
    "text": "h1",
  },
]
`;

exports[`vitePluginMdReTagTest > 从 .md 文本生成 Toc 数据 2`] = `[]`;

exports[`vitePluginMdReTagTest > 从 .md 文本生成 Toc 数据 3`] = `
[
  {
    "children": [
      {
        "children": [],
        "level": 2,
        "text": "安装组件库",
      },
      {
        "children": [],
        "level": 2,
        "text": "注册组件",
      },
    ],
    "level": 1,
    "text": "快速开始",
  },
]
`;
